plugins/debuginfo-install.py

```diff
@@ -79,54 +83,95 @@ def run(self):
+                    self.debugsource_missing.add(str(package))
+
+        if self.debuginfo_missing:
+            logger.info(_("Could not find debuginfo packages for the following packages: %s") %
+                        ", ".join(sorted(self.debuginfo_missing)))
+
+        if self.debugsource_missing:
+            logger.info(_("Could not find debugsource packages for the following packages: %s") %
+                        ", ".join(sorted(self.debugsource_missing)))
 
         if errors_spec and self.base.conf.strict:
             raise dnf.exceptions.PackagesNotAvailableError(_("Unable to find a match"),
                                                            pkg_spec=' '.join(errors_spec))
 
-    def _dbg_available(self, dbgname, package, match_evra):
-        if match_evra:
-            return self.packages_available.filter(
-                name=dbgname,
-                epoch=int(package.epoch),
-                version=str(package.version),
-                release=str(package.release),
-                arch=str(package.arch))
-        else:
-            return self.packages_available.filter(
-                name=dbgname,
-                arch=str(package.arch))
-
-    def _di_install(self, package):
-        for dbgname in [package.debug_name, package.source_debug_name]:
-            if dbgname in self.dbgdone:
-                break
-            if self._dbg_available(dbgname, package, True):
-                di = "{0}-{1}:{2}-{3}.{4}".format(
-                    dbgname,
-                    package.epoch,
-                    package.version,
-                    package.release,
-                    package.arch)
-                self.base.install(di)
-            elif self._dbg_available(dbgname, package, False):
-                di = "{0}.{1}".format(dbgname, package.arch)
-                self.base.install(di)
-            else:
-                continue
-            self.dbgdone.append(dbgname)
-            break
-        else:
-            logger.info(
-                _("Could not find debuginfo for package: %s") % package)
+    def _install(self, package, debug_name, base_nevra):
+        if debug_name in self.packages_installed:
+            return True
+
+        kwargs = {}
+
+        # if some part of EVRA was specified in the argument, add it as a filter
+        if base_nevra.epoch is not None:
+            kwargs["epoch"] = base_nevra.epoch
+        if base_nevra.version is not None:
+            kwargs["version"] = base_nevra.version
+        if base_nevra.release is not None:
+            kwargs["release"] = base_nevra.release
+        if base_nevra.arch is not None:
+            kwargs["arch"] = base_nevra.arch
+
+        # if the package is installed and none of EVR was specified in the
+        # argument, use the installed package's EVR (install debuginfo with the
+        # same version as the installed package)
+        if package._from_system and \
```

Comment:
I believe there is a problem with arch value in case that secondary architecture is installed.
foo.i686 installed
I believe that with argument foo, debuginfo for x86_64 will be installed. I am proposing to add:
```
if package._from_system:
    kwargs["arch"] = package.arch
```

Reply:
Also globs in version, release, arch will make the function non-functional.
Keuwords could be replaced like "release__glob".
I am also not sure, how the if will handle it.
Probably conditions ```                 ("epoch" not in kwargs or kwargs["epoch"] == package.epoch) and \
                ("version" not in kwargs or kwargs["version"] == package.version) and \
                ("release" not in kwargs or kwargs["release"] == package.release): ```
could be removed, because installed package already matched version, release or so on in previous query.

Reply:
> Also globs in version, release, arch will make the function non-functional.
> Keuwords could be replaced like "release__glob".
> I am also not sure, how the if will handle it.

Good observation. I've just tested it, it will not find the correct package when a glob is specified. However, what is a glob supposed to do here? The way it's supposed to work is the user specifies a version of the `-debuginfo` to install. If they specify a glob, which version should it install? Like the highest one matching the glob? I'm not sure we want to support this at all?

> Probably conditions ("epoch" not in kwargs or kwargs["epoch"] == package.epoch) and \ ("version" not in kwargs or kwargs["version"] == package.version) and \ ("release" not in kwargs or kwargs["release"] == package.release):
could be removed, because installed package already matched version, release or so on in previous query.

You're right, I think it can actually be removed and the tests do pass with it gone... :slightly_smiling_face:

Reply:
When glob is used in version i would suggest:
1. In set is installed package use version of installed package
2. Let solver to decide - mostly the highest version
You only need to change kwargs["version"] to kwargs["version__glob"], kwargs["release"] to kwargs["release__glob"], and kwargs["arch"] to kwargs["arch__glob"]. parsing epoch do not support globs.
The rest of the code will handle it correctly.